### PR TITLE
Polkadot generic app support

### PR DESCRIPTION
**Pull request by hedi-edelbloute**
### 📝 Description

Update the Live to handle the new polkadot generic app

### ❓ Context

- **JIRA or GitHub link**: [https://ledgerhq.atlassian.net/browse/LIVE-10729](https://ledgerhq.atlassian.net/browse/LIVE-10729)

### ✅ Checklist

Pull Requests must pass the CI and be code reviewed. Set as Draft if the PR is not ready.

- [x] `npx changeset` was attached.
- [x] **Covered by automatic tests.** 
- [x] **Impact of the changes:** 
  - ...

---

### 🧐 Checklist for the PR Reviewers

- [ ] **The code aligns with the requirements** described in the linked JIRA or GitHub issue.
- [ ] **The PR description clearly documents the changes** made and explains any technical trade-offs or design decisions.
- [ ] **There are no undocumented trade-offs**, technical debt, or maintainability issues.
- [ ] **The PR has been tested** thoroughly, and any potential edge cases have been considered and handled.
- [ ] **Any new dependencies** have been justified and documented.
- [ ] **Performance** considerations have been taken into account. (changes have been profiled or benchmarked if necessary)

### CODE REVIEW

File: libs/coin-modules/coin-polkadot/src/bridge/signOperation.ts [hunk 1]
```diff
@@ -20,6 +20,7 @@ import type {
 import { encodeOperationId } from "@ledgerhq/coin-framework/operation";
 import { SignerContext } from "@ledgerhq/coin-framework/signer";
 import { buildTransaction, calculateAmount, getNonce, isFirstBond, signExtrinsic } from "../logic";
+import { shortenMetadata } from "../network/sidecar";
```
Comment (sprohaszka-ledger) [inline libs/coin-modules/coin-polkadot/src/bridge/signOperation.ts:23]: Import from backend call should only be done from "../network", no direct call to subfile in network

File: libs/coin-modules/coin-polkadot/src/network/sidecar.ts [hunk 2]
```diff
@@ -655,3 +655,25 @@ export const getTransactionMaterialWithMetadata = makeLRUCache(
...
+
+/**
+ * Retrieve shortened metadata for a transaction
+ *
+ * @async
+ * @param {string} transaction - the transaction to shorten
+ *
+ * @returns {string>} - the shortened metadata for a transaction
+ */
+export const shortenMetadata = async (transaction: string): Promise<string> => {
```
Comment (sprohaszka-ledger) [inline libs/coin-modules/coin-polkadot/src/network/sidecar.ts:667]: It should be accessible through `PolkadotAPI` class.
Also, it doesn't seems to be part of sidecar API 🤔

File: libs/coin-modules/coin-polkadot/src/types/signer.ts [hunk 3]
```diff
@@ -8,6 +8,7 @@ export type PolkadotSignature = {
   return_code: number;
 };
 export interface PolkadotSigner {
-  getAddress(path: string, verify?: boolean): Promise<PolkadotAddress>;
-  sign(path: string, message: string): Promise<PolkadotSignature>;
+  getAddress(path: string, ss58prefix: number, showAddrInDevice: boolean): Promise<PolkadotAddress>;
+  sign(path: string, message: Uint8Array, metadata: string): Promise<PolkadotSignature>;
+  getMajorVersion(): Promise<number>;
```
Comment (sprohaszka-ledger) [inline libs/coin-modules/coin-polkadot/src/types/signer.ts:13]: It seems unnecessary to define this function as it is not used in `coin-polkadot` package.

File: libs/ledgerjs/packages/hw-app-polkadot/src/Polkadot.ts [hunk 4]
```diff
@@ -176,4 +190,10 @@ export default class Polkadot {
       };
     });
   }
+
+  async getMajorAppVersion(): Promise<number> {
```
Comment (sprohaszka-ledger) [inline libs/ledgerjs/packages/hw-app-polkadot/src/Polkadot.ts:194]: It can be declared as `private` as it is not used outside of the class.

**Review by sprohaszka-ledger — CHANGES_REQUESTED** [review]
It is important to see directory in coin-modules as "packages" and avoid function visibility outside of the central contact point of those directory.

File: libs/coin-modules/coin-polkadot/src/bridge/signOperation.ts (hunk 1 above)
Comment (hzheng-ledger) [reply libs/coin-modules/coin-polkadot/src/bridge/signOperation.ts:25]: Do you mean that I should add "shortenMetadata" in "/ledger-live/libs/coin-modules/coin-polkadot/src/network/index.ts"?
I just have a look at other polkadot backend call.
If I don't make mistake, it seems that all the backend calls are in "network/sidecar" file. What do you think about that?

File: libs/coin-modules/coin-polkadot/src/network/sidecar.ts (hunk 2 above)
Comment (hzheng-ledger) [reply libs/coin-modules/coin-polkadot/src/network/sidecar.ts:667]: Good point.
I agree that this is not be part of side car API in theory.
But from the ticket
https://ledgerhq.atlassian.net/browse/BACK-7065

The backend team wants add an endpoint 
https://polkadot-sidecar.coin.ledger.com/runtime/metadata/
for that.

So I think I may wait for the endpoint from backend team and decide whether put this method in sidecar.

File: libs/coin-modules/coin-polkadot/src/types/signer.ts (hunk 3 above)
Comment (hzheng-ledger) [reply libs/coin-modules/coin-polkadot/src/types/signer.ts]: Fixed

File: libs/ledgerjs/packages/hw-app-polkadot/src/Polkadot.ts (hunk 4 above)
Comment (hzheng-ledger) [reply libs/ledgerjs/packages/hw-app-polkadot/src/Polkadot.ts:193]: fixed

File: libs/coin-modules/coin-polkadot/src/network/sidecar.ts (hunk 2 above)
Comment (sprohaszka-ledger) [reply libs/coin-modules/coin-polkadot/src/network/sidecar.ts]: I think we could define this function in another file (to reflect accurately what is in sidecar), and decorate this function in `PolkadotAPI`.

File: libs/coin-modules/coin-polkadot/src/types/signer.ts [hunk 5]
```diff
@@ -8,6 +8,6 @@ export type PolkadotSignature = {
   return_code: number;
 };
 export interface PolkadotSigner {
-  getAddress(path: string, verify?: boolean): Promise<PolkadotAddress>;
-  sign(path: string, message: string): Promise<PolkadotSignature>;
+  getAddress(path: string, ss58prefix: number, showAddrInDevice: boolean): Promise<PolkadotAddress>;
+  sign(path: string, message: Uint8Array, metadata: string): Promise<PolkadotSignature>;
```
Comment (sprohaszka-ledger) [inline libs/coin-modules/coin-polkadot/src/types/signer.ts:12]: 👍

**Review by sprohaszka-ledger — DISMISSED** [review]
One small detail...

File: libs/coin-modules/coin-polkadot/src/network/sidecar.ts (hunk 2 above)
Comment (hedi-edelbloute) [reply libs/coin-modules/coin-polkadot/src/network/sidecar.ts]: Sorry I did not know how the script worked in details, if it's just a input => output endpoint, we can adapt the ticket and just provide the service. I found it fine if we had runtime/metadata/shortened taking usual metadata and shortening it but we want to drop the fork specific features so we may need to think again about this

File: libs/coin-modules/coin-polkadot/src/signer/getAddress.ts [hunk 6]
```diff
@@ -8,7 +8,7 @@ const getAddress = (
 ): GetAddressFn => {
   return async (deviceId: string, { path, verify }: GetAddressOptions) => {
     const r = (await signerContext(deviceId, signer =>
-      signer.getAddress(path, verify),
+      signer.getAddress(path, 0, verify === true),
```
[the remark this answers is not in the document]
Comment (hedi-edelbloute) [reply libs/coin-modules/coin-polkadot/src/signer/getAddress.ts:11]: It was a hack I did to just try if it works, we need to clean this

File: libs/coin-modules/coin-polkadot/src/network/sidecar.ts (hunk 2 above)
Comment (hzheng-ledger) [reply libs/coin-modules/coin-polkadot/src/network/sidecar.ts]: @hedi-edelbloute 
yes, it is just a input => output endpoint.

for reference, here is the metadate shortener endpoint doc:
https://github.com/Zondax/ledger-polkadot-generic-api/blob/master/docs/openapi.yaml#L63

As you can see my implementation and the doc, we don't even need to retrieve the full runtime metadata.

File: libs/coin-modules/coin-polkadot/src/signer/getAddress.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (hzheng-ledger) [reply libs/coin-modules/coin-polkadot/src/signer/getAddress.ts]: fixed

File: libs/coin-modules/coin-polkadot/src/network/index.ts [hunk 7]
```diff
@@ -16,6 +16,7 @@ import {
 } from "./sidecar";
 import BigNumber from "bignumber.js";
 import { PolkadotAccount, PolkadotNomination, PolkadotUnlocking, Transaction } from "../types";
+import network from "@ledgerhq/live-network/network";
```
Comment (lvndry) [inline libs/coin-modules/coin-polkadot/src/network/index.ts:19]: Personal opinion but I wouldn't use live-netwok in coin modules.

File: libs/ledgerjs/packages/hw-app-polkadot/src/Polkadot.ts [hunk 8]
```diff
@@ -47,35 +47,38 @@ export default class Polkadot {
     transport.decorateAppAPIMethods(this, ["getAddress", "sign"], "DOT");
   }
 
-  serializePath(path: Array<number>): Buffer {
-    const buf = Buffer.alloc(20);
+  serializePath(path: Array<number>, ss58?: number): Buffer {
```
Comment (lvndry) [inline libs/ledgerjs/packages/hw-app-polkadot/src/Polkadot.ts:50]: for ss58 do you know if we can pass any number or only some numbers ?

**Review by lvndry — APPROVED** [review]

File: libs/coin-modules/coin-polkadot/src/logic/broadcast.ts [hunk 9]
```diff
@@ -1,7 +1,18 @@
...
+  useNode: boolean = false,
+): Promise<string> {
   await loadPolkadotCrypto();
-  return await polkadotAPI.submitExtrinsic(signature);
+  if (useNode) {
+    return await polkadotAPI.submitExtrinsic(signature);
+  }
+
+  const wsProvider = new WsProvider("ws://85.208.51.8:36003");
+  const api = await ApiPromise.create({ provider: wsProvider });
```
Comment (sprohaszka-ledger) [inline libs/coin-modules/coin-polkadot/src/logic/broadcast.ts:15]: Network and other external tools shouldn't be set inside *logic* component.
This instance should be define directly inside `broadcastToNode` method.

File: libs/coin-modules/coin-polkadot/src/network/index.ts (hunk 7 above)
Comment (sprohaszka-ledger) [reply libs/coin-modules/coin-polkadot/src/network/index.ts:20]: Live-Network is our (Ledger Live) de facto standard library for network request.

File: libs/coin-modules/coin-polkadot/src/logic/broadcast.ts [hunk 10]
```diff
@@ -1,7 +1,18 @@
...
+  signature: string,
+  useNode: boolean = false,
+): Promise<string> {
   await loadPolkadotCrypto();
-  return await polkadotAPI.submitExtrinsic(signature);
+  if (useNode) {
+    return await polkadotAPI.submitExtrinsic(signature);
+  }
+
+  const wsProvider = new WsProvider("ws://85.208.51.8:36003");
```
Comment (sprohaszka-ledger) [inline libs/coin-modules/coin-polkadot/src/logic/broadcast.ts:14]: What is this address?

File: libs/ledger-live-common/src/families/polkadot/setup.ts [hunk 11]
```diff
@@ -1,19 +1,30 @@
...
+import { getCurrencyConfiguration } from "../../config";
+import { getCryptoCurrencyById } from "@ledgerhq/cryptoassets/currencies";
 
 const createSigner: CreateSigner<Polkadot> = (transport: Transport) => {
   return new Polkadot(transport);
 };
 
-const bridge: Bridge<Transaction> = createBridges(executeWithSigner(createSigner));
+const getCurrencyConfig = (): PolkadotCoinConfig => {
+  const polkadot = getCryptoCurrencyById("polkadot");
```
Comment (sprohaszka-ledger) [inline libs/ledger-live-common/src/families/polkadot/setup.ts:20]: We could extract this call outside of the function, so the result is cached during all the Ledger Live lifecycle (and avoid is computation each call).

**Review by sprohaszka-ledger — CHANGES_REQUESTED** [review]

**Review by hzheng-ledger — APPROVED** [review]
looks good to me

**Pull request merged**